File: Data Structures & Algorithms/two-integer-sum/submission-4.cpp
// 1. Turn off I/O sync
auto init = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    return 0;
}();

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> x;
        // 2. Prevent rehashing
        x.reserve(nums.size()); 
        
        for (int i = 0; i < nums.size(); i++) {
            int item = nums[i];
            int complement = target - item;
            
            // 3. Single lookup using the iterator
            auto it = x.find(complement);
            if (it != x.end()) {
                return {it->second, i};
            }
            
            x[item] = i;
        }
        
        return {};
    }
};

File: Data Structures & Algorithms/binary-search/submission-4.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        mid = 0
        while left < right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                right = mid 
            else:
                left = mid + 1
        
        return -1 if nums[left] != target else left